ledger/participant-integration-api/src/main/scala/platform/store/cache/EventsBuffer.scala

```diff
@@ -0,0 +1,167 @@
+  *
+  * @param maxBufferSize The maximum buffer size.
+  * @param metrics The DAML metrics.
+  * @param bufferQualifier The qualifier used for metrics tag specialization.
+  * @param isRangeEndMarker Identifies if an element [[E]] should be treated
+  *                         as a range end marker.
+  * @tparam O The offset type.
+  * @tparam E The entry buffer type.
+  */
+private[cache] final class EventsBuffer[O: Ordering, E](
+    maxBufferSize: Int,
+    metrics: Metrics,
+    bufferQualifier: String,
+    isRangeEndMarker: E => Boolean,
+) {
+  @volatile private var bufferRef = Vector.empty[(O, E)]
+
+  private var bufferRangeEnd: Option[O] = None
+
+  private val pushTimer = metrics.daml.services.index.buffer.push(bufferQualifier)
+  private val sliceTimer = metrics.daml.services.index.buffer.slice(bufferQualifier)
+  private val pruneTimer = metrics.daml.services.index.buffer.prune(bufferQualifier)
+
+  /** Appends a new event to the buffer.
+    *
+    * Starts evicting from the tail when `maxBufferSize` is reached.
+    *
+    * @param offset The event offset.
+    *              Must be higher than the last appended entry's offset, with the exception
+    *              of the range end marker, which can have an offset equal to the last appended element.
+    * @param entry The buffer entry.
+    */
+  def push(offset: O, entry: E): Unit =
+    Timed.value(
+      pushTimer,
+      synchronized {
+        if (bufferRef.size == maxBufferSize)
+          bufferRef = bufferRef.drop(1)
+
+        bufferRangeEnd.foreach { lastOffset =>
+          // Ensure vector grows with strictly monotonic offsets.
+          // Only specially-designated range end markers are allowed
+          // to have offsets equal to the buffer range end.
+          if (lastOffset > offset || (lastOffset == offset && !isRangeEndMarker(entry))) {
+            throw UnorderedException(lastOffset, offset)
+          }
+        }
+
+        // Update the buffer range end
+        bufferRangeEnd = Some(offset)
+
+        if (!isRangeEndMarker(entry)) {
+          // The range end markers are not appended to the buffer
+          bufferRef = bufferRef :+ offset -> entry
+        }
+      },
+    )
+
+  /** Returns a slice of events from the buffer.
+    *
+    * Throws an exception if requested with `endInclusive` higher than the highest offset in the buffer.
+    *
+    * @param startExclusive The start exclusive bound of the requested range.
+    * @param endInclusive The end inclusive bound of the requested range.
+    * @return The series of events as an ordered vector satisfying the input bounds.
+    */
+  def slice(startExclusive: O, endInclusive: O): Vector[(O, E)] =
+    Timed.value(
+      sliceTimer, {
+        val buffer = bufferRef
+        if (buffer.isEmpty) {
+          Vector.empty
+        } else if (bufferRangeEnd.exists(_ < endInclusive)) {
+          throw RequestOffBufferBounds(buffer.last._1, endInclusive)
```

Comment:
Shouldn't the condition on `endInclusive` not being greater than `bufferRangeEnd` be before returning an empty buffer to have consistent behavior?
E.g. after pruning the whole buffer and asking for an out-of-range end offset we won't get an exception but an empty buffer.

Reply:
You are right. For the sake of consistency, they should be inverted.